static/js/lib/coupon.js

```diff
@@ -29,34 +30,81 @@ function makeProgramIdLookup<T: HasProgramId>(arr: Array<T>): Map<number, T> {
   return new Map(arr.map((value: T) => [value.program_id, value]));
 }
 
-function* genPrices(programs: Array<Program>, prices: CoursePrices, coupons: Coupons) {
+// This calculates coupon adjusted prices (and unadjusted prices) for all programs, courses, and course runs.
+export const calculatePrices = (
+  programs: Array<Program>, prices: CoursePrices, coupons: Coupons
+): CouponPrices => {
   const couponLookup: Map<number, Coupon> = makeProgramIdLookup(coupons);
   const priceLookup: Map<number, CoursePrice> = makeProgramIdLookup(prices);
 
+  const pricesInclCouponByRun = new Map();
+  const pricesInclCouponByCourse = new Map();
+  const pricesInclCouponByProgram = new Map();
+  const pricesExclCouponByProgram = new Map();
+
   for (const program of programs) {
+    let priceObj = priceLookup.get(program.id);
+    if (!priceObj) {
+      // Shouldn't get here, we should only be calling this function
+      // if we retrieved all the values from the API already
+      throw 'Unable to find program to get the price';
+    }
+    let originalPrice = priceObj.price;
+    // Currently only one coupon per program is allowed, even if that program only affects one course
+    let coupon = couponLookup.get(program.id);
+    // Note that this variable is independent of what the coupon is for
+    let adjustedPrice = originalPrice;
+    if (coupon) {
+      adjustedPrice = calculateDiscount(originalPrice, coupon.amount_type, coupon.amount);
+    }
+
+    let priceExclCouponByProgram = {
+      price: originalPrice,
+      coupon: null,
+    };
+    let isProgramCoupon = (
```

Comment:
I'm not sure about that link, we don't have a `obj` to use with `isProgramCoupon`